rclpy_components/rclpy_components/component_manager.py

```diff
@@ -0,0 +1,120 @@
+        super().__init__(name, **kwargs)
+        self.executor = executor
+        # Implement the 3 services described in
+        # http://design.ros2.org/articles/roslaunch.html#command-line-arguments
+        self.list_node_srv_ = self.create_service(ListNodes, "~/_container/list_nodes", self.on_list_node)
+        self.load_node_srv_ = self.create_service(LoadNode, "~/_container/load_node", self.on_load_node)
+        self.unload_node_srv_ = self.create_service(UnloadNode, "~/_container/unload_node", self.on_unload_node)
+
+        self.components = {}  # key: unique_id, value: full node name and component instance
+        self.unique_id_index = 0
+
+        self.executor.spin()
+
+    def gen_unique_id(self):
+        self.unique_id_index += 1
+        return self.unique_id_index
+
+    def on_list_node(self, req: ListNodes.Request, res: ListNodes.Response):
+        res.unique_ids = [int(key) for key in self.components.keys()]
+        res.full_node_names = [v[0] for v in self.components.values()]
+
+        return res
+
+    def on_load_node(self, req: LoadNode.Request, res: LoadNode.Response):
+        component_entry_points = entry_points().get(RCLPY_COMPONENTS, None)
+        if not component_entry_points:
+            logger.error('No rclpy components registered')
+            res.success = False
+            return res
+
+        component_entry_point = None
+        for ep in component_entry_points:
+            if ep.name == req.plugin_name:
+                component_entry_point = ep
+                break
+
+        if not component_entry_point:
+            logger.error('No rclpy component found by %s' % req.plugin_name)
+            res.success = False
+            return res
+
+        component_class = component_entry_point.load()
+        node_name = req.node_name if req.node_name else \
+            str.lower(str.split(component_entry_point.value, ':')[1])
+
+        params_dict = {'use_global_arguments': False}
+        if req.parameters:
+            params_dict['parameter_overrides'] = req.parameters
+
+        if req.node_namespace:
+            params_dict['namespace'] = req.node_namespace
+
+        if req.remap_rules:
+            params_dict['cli_args'] = ['--ros-args']
+            for rule in req.remap_rules:
+                params_dict['cli_args'].extend(['-r', rule])
+
+        try:
+            logger.info('Instantiating {} with {}, {}'.format(component_entry_point.value, node_name, params_dict))
+            component = component_class(node_name, **params_dict)
+
+            res.unique_id = self.gen_unique_id()
+            # TODO Assign the full_node_name with node.get_fully_qualified_name
+            res.full_node_name = '/{}'.format(node_name)
+            if req.node_namespace:
+                res.full_node_name = '/{}{}'.format(req.node_namespace, res.full_node_name)
+            res.success = True
+            self.components[str(res.unique_id)] = (res.full_node_name, component)
```

Comment:
Why not just use the integer value for the key? Then we don't have to cast between types.